keylime/cert_utils.py

```diff
@@ -1,13 +1,133 @@
+    """
+    try:
+        ek509 = x509_der_cert(ekcert)
+
+        trusted_certs = tpm_ek_ca.cert_loader(tpm_cert_store)
+        for cert_file, pem_cert in trusted_certs.items():
+            signcert = x509_pem_cert(pem_cert)
+            if ek509.issuer != signcert.subject:
+                continue
+
+            signcert_pubkey = signcert.public_key()
+            try:
+                if isinstance(signcert_pubkey, RSAPublicKey):
+                    signcert_pubkey.verify(
+                        ek509.signature,
+                        ek509.tbs_certificate_bytes,
+                        padding.PKCS1v15(),
+                        ek509.signature_hash_algorithm,
+                    )
+                elif isinstance(signcert_pubkey, EllipticCurvePublicKey):
+                    signcert_pubkey.verify(
+                        ek509.signature,
+                        ek509.tbs_certificate_bytes,
+                        ec.ECDSA(ek509.signature_hash_algorithm),
+                    )
+                else:
+                    logger.warning("Unsupported public key type: %s", type(signcert_pubkey))
+                    continue
+            except crypto_exceptions.InvalidSignature:
+                continue
+
+            logger.debug("EK cert matched cert: %s", cert_file)
+            return True
+    except Exception as e:
+        # Log the exception so we don't lose the raw message
+        logger.exception(e)
+        raise Exception("Error processing ek/ekcert. Does this TPM have a valid EK?").with_traceback(sys.exc_info()[2])
+
+    logger.error("No Root CA matched EK Certificate")
+    return False
+
+
+def verify_ek_script(script, env, cwd):
+    if script is None:
+        logger.warning("External check script (%s) not specified", script)
+        return False
+
+    script_path = os.path.abspath(script)
+    if not os.path.isfile(script_path):
+        if cwd is None or not os.path.isfile(os.path.abspath(os.path.join(cwd, script))):
+            logger.warning("External check script (%s) not found; please make sure its path is correct", script)
+            return False
+        script_path = os.path.abspath(os.path.join(cwd, script))
+
+    try:
+        proc = subprocess.run(
+            [script_path],
+            env=env,
+            shell=True,
+            cwd=cwd,
+            stderr=subprocess.STDOUT,
+            stdout=subprocess.PIPE,
+            check=False,
+        )
+        if proc.returncode != 0:
+            errmsg = ""
+            if proc.stdout is not None:
+                errmsg = proc.stdout.decode("utf-8")
+            logger.info("External check script failed to validate EK: %s", errmsg)
+            return False
+        logger.debug("External check script successfully to validated EK")
+        if proc.stdout is not None:
+            logger.info("ek_check output: %s", proc.stdout.decode("utf-8"))
+    except Exception as e:
```

Comment:
there should be an specific exception to indicate an error with the subproces

Reply:
Done, please re-check.